### ISSUE TITLE
Service scripts - a new kind of extension scripts

### ISSUE THREAD

**Issue by hugbug** [nzbget/nzbget#318]
In #304 (Minutely Scheduled Tasks) a need for scripts executed regularly and often has raised. A similar use case is in forum topic [Switch to next download in queue on low speed](http://forum.nzbget.net/viewtopic.php?f=3&t=2739).

A workaround was discussed in the linked issue. However to better cover the use case and to simplify the development of such scripts a new kind of extension scripts can be implemented - **service-scripts**.

Service scripts will be defined via new option **ServiceScript** and will have a new signature:
```shell
### NZBGET SERVICE SCRIPT
```

What makes service scripts special:
- service scripts run parallel to NZBGet;
- when NZBGet starts it launches the service scripts;
- service scripts run endless organizing an internal loop;
- if a service script terminates that's treated by NZBGet as an unexpected behavior (e. g. crash) and the script is started again;
- when NZBGet shuts down it sends a termination signal to service scripts, which should gracefully terminate.

So service scripts are similar to scheduler scripts because they are for periodical tasks. The main advantage of service script is that they don't require boot time for each iteration making them more efficient.

A non-efficiency of queue-scripts (such as FakeDetector) makes me think about extending service-scripts with event handling functionality but that's something to implement later.

**Comment by caronc** [comment]
Or you could always just add an option to the Scheduler (when creating a Task) with a radio button: "_Start with NZBGet_" which will effectively do your "__*__" feature.

I'm all for amalgamating the script types into less oppose to creating more types.

The only problem with the current proposed approach is you're duplicating a lot of functionality you've already accomplished in ticket #304. No?

**Comment by hugbug** [comment]
>I'm all for amalgamating the script types into less oppose to creating more types.

Absolutely. That's why #319 (Unified extension scripts settings).

Selecting a script in "Extensions" is easier than configuring a new Task. Not a big difference for you and me but users often complain about complexity of NZBGet configuration. I'm addressing this with "Unified extension scripts settings". Since I can't integrate scheduler scripts there, the idea is to provide a scheduler-like alternative which doesn't require configuring of a scheduler task. Plus an improvement of efficiency.

I think most scheduler scripts can be converted to service scripts with advantages.

>The only problem with the current proposed approach is you're duplicating a lot of functionality you've already accomplished in ticket #304. No?

I'll find a way to reuse it :wink:

At the end this issue is a proposal, which I wanted to discuss with you and other developers. I haven't implemented anything yet. Thanks for your feedback!

**Comment by caronc** [comment]
Fair enough;  I just worry about backwards compatibility.  The one thing i love about your (already implemented) solution (in #304) is that it addresses this _exact_ issue already. Your solution (in #304)  already allows anyone using a schedule script to continue doing so as well as having it run automatically at start up. I just wouldn't want you to 'undo' this because it gracefully even supports backwards compatibility for those who want to run indefinitely and those who don't!

What you're proposing here is just duplication of code that will 'only' work with people who update to this version or newer.

**For the record**: I love where you're going with #319! Maybe there is a solution from that direction instead? Just please, please, please.... no more unique Script types!! :)

**Comment by hugbug** [comment]
Don't worry, I'm not going to undo #304.

> I love where you're going with #319!

That's the point. In particular setting of scheduler scripts is currently complicated. I want to make activating of DirWatch script as simple as selecting it in "Extensions".

> What you're proposing here is just duplication of code

I don't know what you mean with duplication. The graceful termination applies to all script types and will not be written again. The automatic script restart (on crash) doesn't exist yet.

> code that will 'only' work with people who update to this version or newer.

He-he, you can't stop progress 😉. Updating NZBGet is simple, everyone should use the latest version.

> Just please, please, please.... no more unique Script types!! :)

It's hard to add new functionality without changing anything. 

All current script types are meant to run-and-terminate on a given event. What is proposed in this issue is a totally different technique of running scripts in parallel to NZBGet. So this is **THE** script type that has potential to replace all others in the future.

> I just worry about backwards compatibility.

Backwards compatibility is a top priority. The "Unified script settings" makes it appear for users as a new extension concept although keeping the complete compatibility with existing scripts. If I could redesign the extension concept I would create only one script type with extendable events (similar to queue scripts).

> no more unique Script types!! :)

First of all, there will no be a separate option ServiceScript as it will be covered by option "Extensions". The separate script signature is important only for script authors. One way or another they need to know details. Writing service script should be easier than misusing scheduler script for "always run in parallel". 

Nonetheless if you need (want) compatibility with older NZBGet versions you still can use the scheduler script technique. You can make your script multi-type script (`NZBGET SCHEDULER/SERVICE SCRIPT`), which will allow easy config in v18 but still compatible with older versions (with more config in scheduler).

Although it seems possible to misuse scheduler scripts it doesn't look like a better alternative. Let's see. I can't show all scheduler scripts selectable in "Extensions" because only scheduler scripts written in a certain way should be shown there. These scripts need a special marker. How is such a marker better than the "SERVICE SCRIPT" signature?

When I'm talking about potential I mean the merging of queue-scripts and service-scripts. Ideally an always-running-service-script should be notified about queue events. This will make scripts like FakeDetector much more efficient. And the small latency opens potential for other event types.

Thanks for your feedback.

**Comment by caronc** [comment]
Thanks for your detailed response; you addressed my key points.

If you did something like **_AUTO_ SERVICE SCRIPT ###** to auto-start the service script on boot and just **SERVICE SCRIPT ###** to manually add it in the Services area?  Not sure how that would play with backwards compatibility? It's pretty much what you're already eluding to though with your _SERVICE ##_ reference.

> Nonetheless if you need (want) compatibility with older NZBGet versions you still can use the scheduler script technique. You can make your script multi-type script (NZBGET SCHEDULER/SERVICE SCRIPT), which will allow easy config in v18 but still compatible with older versions (with more config in scheduler).

That's cool! I didn't know it would still work that way (with earlier versions).

> I don't know what you mean with duplication. The graceful termination applies to all script types and will not be written again. The automatic script restart (on crash) doesn't exist yet.

I'll just expand on what I mean here. The SCHEDULER option (with respect to NZBGet v18) can:
*  :white_check_mark: Automatically run a script on scheduled intervals.
*  :white_check_mark: :key: Automatically run a script when NZBGet starts up.

The SERVICE Script option you're proposing (here in this ticket):
*   :white_check_mark: :key: Automatically run a script when NZBGet starts up.

Either way; thanks for listening to me rant! Keep up the awesome work! :smile:

**Comment by hugbug** [comment]
You convinced me.
See alternative solution in [how to make scheduler scripts selectable in option "Extensions"](https://github.com/nzbget/nzbget/issues/319#issuecomment-268363155).

**The "Service scripts"-feature will not be implemented, at least not until it can replace queue scripts.**
